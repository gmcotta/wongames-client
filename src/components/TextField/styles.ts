import styled, { css, DefaultTheme } from 'styled-components'
import { IconPositionOptions } from '.'

type WrapperProps = {
  disabled?: boolean
  hasError?: boolean
}

const wrapperModifiers = {
  disabled: (theme: DefaultTheme) => css`
    ${Label},
    ${InputWrapper},
    ${Input} {
      cursor: not-allowed;
      color: ${theme.colors.grayDisabled};
      svg {
        color: ${theme.colors.grayDisabled};
      }
      &::placeholder {
        color: ${theme.colors.grayDisabled};
      }
    }
  `,
  hasError: (theme: DefaultTheme) => css`
    ${Label},
    ${InputWrapper},
    ${Input} {
      color: ${theme.colors.red};
      svg {
        color: ${theme.colors.red};
      }
      &::placeholder {
        color: ${theme.colors.red};
      }
    }

    ${InputWrapper} {
      border-color: ${theme.colors.red};
    }
  `
}

export const Wrapper = styled.div<WrapperProps>`
  ${({ theme, disabled, hasError }) => css`
    ${disabled && wrapperModifiers.disabled(theme)}
    ${hasError && wrapperModifiers.hasError(theme)}
  `}
`

export const Label = styled.label`
  ${({ theme }) => css`
    display: inline-flex;
    font-size: ${theme.font.sizes.small};
    color: ${theme.colors.black};
    cursor: pointer;
    padding-bottom: calc(${theme.spacings.xxsmall} / 2);
  `}
`

type InputWrapperProps = {
  hasIcon?: boolean
  iconPosition?: IconPositionOptions
  disabled?: boolean
}

const inputWrapperModifiers = {
  hasIcon: (theme: DefaultTheme) => css`
    svg {
      width: 2.4rem;
      color: ${theme.colors.darkGray};
    }
  `,
  iconPosition: (theme: DefaultTheme, iconPosition: IconPositionOptions) => css`
    svg {
      ${iconPosition === 'left' &&
      css`
        margin-right: ${theme.spacings.xsmall};
      `}
      ${iconPosition === 'right' &&
      css`
        margin-left: ${theme.spacings.xsmall};
      `}
    }
  `
}

export const InputWrapper = styled.div<InputWrapperProps>`
  ${({ theme, hasIcon, iconPosition }) => css`
    display: flex;
    background: ${theme.colors.lightGray};
    border-radius: 0.2rem;
    padding: 0 ${theme.spacings.xsmall};
    border: 0.2rem solid ${theme.colors.lightGray};
    height: 5rem;
    ${hasIcon && inputWrapperModifiers.hasIcon(theme)}
    ${iconPosition && inputWrapperModifiers.iconPosition(theme, iconPosition)}

    &:focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }
  `}
`

export const Input = styled.input`
  ${({ theme }) => css`
    color: ${theme.colors.black};
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} 0;
    background: transparent;
    border: 0;
    outline: none;
    width: 100%;

    &:-webkit-autofill {
      -webkit-box-shadow: 0 0 0 ${theme.spacings.small}
        ${theme.colors.lightGray} inset;
      filter: none;
    }
  `}
`

export const ErrorMessage = styled.span`
  ${({ theme }) => css`
    color: ${theme.colors.red};
    font-size: ${theme.font.sizes.xsmall};
  `}
`
