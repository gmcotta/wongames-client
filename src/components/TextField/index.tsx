import { InputHTMLAttributes, useState, ChangeEvent, ReactNode } from 'react'
import * as S from './styles'

export type IconPositionOptions = 'left' | 'right'

export type TextFieldProps = {
  onInput?: (value: string) => void
  label?: string
  labelFor?: string
  initialValue?: string
  icon?: ReactNode
  iconPosition?: IconPositionOptions
  disabled?: boolean
  errorMessage?: string
} & InputHTMLAttributes<HTMLInputElement>

const TextField = ({
  onInput,
  label,
  labelFor = '',
  initialValue = '',
  icon,
  iconPosition = 'left',
  disabled = false,
  errorMessage = '',
  ...props
}: TextFieldProps) => {
  const [value, setValue] = useState(initialValue)
  const onChange = (evt: ChangeEvent<HTMLInputElement>) => {
    const newValue = evt.target.value
    setValue(newValue)

    !!onInput && onInput(newValue)
  }
  return (
    <S.Wrapper disabled={disabled} hasError={!!errorMessage}>
      {!!label && <S.Label htmlFor={labelFor}>{label}</S.Label>}
      <S.InputWrapper hasIcon={!!icon} iconPosition={iconPosition}>
        {iconPosition === 'left' && icon}
        <S.Input
          type="text"
          value={value}
          onChange={onChange}
          {...props}
          disabled={disabled}
        />
        {iconPosition === 'right' && icon}
      </S.InputWrapper>
      {!!errorMessage && <S.ErrorMessage>{errorMessage}</S.ErrorMessage>}
    </S.Wrapper>
  )
}

export default TextField
