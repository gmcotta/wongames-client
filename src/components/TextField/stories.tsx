import { Story, Meta } from '@storybook/react/types-6-0'
import { Email } from '@styled-icons/material-outlined'
import TextField, { TextFieldProps } from '.'

export default {
  title: 'Form/TextField',
  component: TextField,
  args: {
    label: 'E-mail',
    labelFor: 'Email',
    id: 'Email',
    initialValue: '',
    placeholder: 'john.doe@gmail.com',
    iconPosition: 'left',
    disabled: false
  },
  argTypes: {
    onInput: { action: 'changed' },
    icon: { type: '' }
  }
} as Meta

export const Default: Story<TextFieldProps> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TextField {...args} />
  </div>
)

export const WithIcon: Story<TextFieldProps> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TextField {...args} icon={<Email />} />
  </div>
)

export const WithErrorMessage: Story<TextFieldProps> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TextField
      {...args}
      icon={<Email />}
      errorMessage="Something is wrong..."
    />
  </div>
)
