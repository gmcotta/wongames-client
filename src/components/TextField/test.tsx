import { Email } from '@styled-icons/material-outlined'
import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { renderWithTheme } from 'utils/tests/helpers'

import TextField from '.'

describe('<TextField />', () => {
  it('should render with label', () => {
    renderWithTheme(<TextField label="label" labelFor="field" id="field" />)

    expect(screen.getByLabelText(/label/i)).toBeInTheDocument()
  })

  it('should render without label', () => {
    renderWithTheme(<TextField id="field" />)

    expect(screen.queryByLabelText(/label/i)).not.toBeInTheDocument()
  })

  it('should render with placeholder', () => {
    renderWithTheme(<TextField id="field" placeholder="example" />)

    expect(screen.getByPlaceholderText(/example/i)).toBeInTheDocument()
  })

  it('should change values when typing', async () => {
    const onInput = jest.fn()
    renderWithTheme(
      <TextField label="label" labelFor="field" id="field" onInput={onInput} />
    )
    const input = screen.getByLabelText(/label/i)
    const text = 'hey you'
    userEvent.type(input, text)
    await waitFor(() => {
      expect(input).toHaveValue(text)
      expect(onInput).toHaveBeenCalledTimes(text.length)
    })
    expect(onInput).toHaveBeenCalledWith(text)
  })

  it('should be accessible by tab', () => {
    renderWithTheme(<TextField label="label" labelFor="field" id="field" />)
    expect(document.body).toHaveFocus()
    const input = screen.getByLabelText(/label/i)
    userEvent.tab()
    expect(input).toHaveFocus()
  })

  it('should render with icon on left side', () => {
    renderWithTheme(
      <TextField
        label="label"
        labelFor="field"
        id="field"
        icon={<Email data-testid="icon" />}
      />
    )
    expect(screen.getByLabelText(/label/i)).toBeInTheDocument()
    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })

  it('should render with icon on right side', () => {
    renderWithTheme(
      <TextField
        label="label"
        labelFor="field"
        id="field"
        icon={<Email data-testid="icon" />}
        iconPosition="right"
      />
    )
    expect(screen.getByLabelText(/label/i)).toBeInTheDocument()
    expect(screen.getByTestId('icon')).toBeInTheDocument()
    expect(screen.getByTestId('icon')).toHaveStyle({
      'margin-left': '1.6rem'
    })
  })

  it('should render with disabled property', () => {
    const { container } = renderWithTheme(
      <TextField label="label" labelFor="field" id="field" disabled />
    )
    expect(screen.getByLabelText(/label/i)).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should not change value when it is disabled', async () => {
    const onInput = jest.fn()
    renderWithTheme(
      <TextField
        label="label"
        labelFor="field"
        id="field"
        disabled
        onInput={onInput}
      />
    )
    const input = screen.getByLabelText(/label/i)
    const text = 'hey you'
    userEvent.type(input, text)
    await waitFor(() => {
      expect(input).not.toHaveValue(text)
      expect(onInput).not.toHaveBeenCalled()
    })
  })

  it('should not be accessible by tab when it is disabled', () => {
    renderWithTheme(
      <TextField label="label" labelFor="field" id="field" disabled />
    )
    expect(document.body).toHaveFocus()
    const input = screen.getByLabelText(/label/i)
    userEvent.tab()
    expect(input).not.toHaveFocus()
  })
})
