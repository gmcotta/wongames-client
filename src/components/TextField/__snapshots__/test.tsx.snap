// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> should render the error message 1`] = `
.c0 .c1,
.c0 .c3,
.c0 .c5 {
  color: #FF6347;
}

.c0 .c1 svg,
.c0 .c3 svg,
.c0 .c5 svg {
  color: #FF6347;
}

.c0 .c1::-webkit-input-placeholder,
.c0 .c3::-webkit-input-placeholder,
.c0 .c5::-webkit-input-placeholder {
  color: #FF6347;
}

.c0 .c1::-moz-placeholder,
.c0 .c3::-moz-placeholder,
.c0 .c5::-moz-placeholder {
  color: #FF6347;
}

.c0 .c1:-ms-input-placeholder,
.c0 .c3:-ms-input-placeholder,
.c0 .c5:-ms-input-placeholder {
  color: #FF6347;
}

.c0 .c1::placeholder,
.c0 .c3::placeholder,
.c0 .c5::placeholder {
  color: #FF6347;
}

.c0 .c3 {
  border-color: #FF6347;
}

.c2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: 1.4rem;
  color: #030517;
  cursor: pointer;
  padding-bottom: calc(0.8rem / 2);
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: #EAEAEA;
  border-radius: 0.2rem;
  padding: 0 1.6rem;
  border: 0.2rem solid #EAEAEA;
}

.c4 svg {
  margin-right: 1.6rem;
}

.c4:focus-within {
  box-shadow: 0 0 0.5rem #F231A5;
}

.c6 {
  color: #030517;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  padding: 0.8rem 0;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
}

.c7 {
  color: #FF6347;
  font-size: 1.2rem;
}

<div
  class="c0"
>
  <label
    class="c1 c2"
    for="field"
  >
    label
  </label>
  <div
    class="c3 c4"
  >
    <input
      class="c5 c6"
      id="field"
      type="text"
      value=""
    />
  </div>
  <span
    class="c7"
  >
    Error message
  </span>
</div>
`

exports[`<TextField /> should render with disabled property 1`] = `
.c0 .c1,
.c0 .c3,
.c0 .c5 {
  cursor: not-allowed;
  color: #B5B5B5;
}

.c0 .c1 svg,
.c0 .c3 svg,
.c0 .c5 svg {
  color: #B5B5B5;
}

.c0 .c1::-webkit-input-placeholder,
.c0 .c3::-webkit-input-placeholder,
.c0 .c5::-webkit-input-placeholder {
  color: #B5B5B5;
}

.c0 .c1::-moz-placeholder,
.c0 .c3::-moz-placeholder,
.c0 .c5::-moz-placeholder {
  color: #B5B5B5;
}

.c0 .c1:-ms-input-placeholder,
.c0 .c3:-ms-input-placeholder,
.c0 .c5:-ms-input-placeholder {
  color: #B5B5B5;
}

.c0 .c1::placeholder,
.c0 .c3::placeholder,
.c0 .c5::placeholder {
  color: #B5B5B5;
}

.c2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: 1.4rem;
  color: #030517;
  cursor: pointer;
  padding-bottom: calc(0.8rem / 2);
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: #EAEAEA;
  border-radius: 0.2rem;
  padding: 0 1.6rem;
  border: 0.2rem solid #EAEAEA;
}

.c4 svg {
  margin-right: 1.6rem;
}

.c4:focus-within {
  box-shadow: 0 0 0.5rem #F231A5;
}

.c6 {
  color: #030517;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  padding: 0.8rem 0;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
}

<div
  class="c0"
  disabled=""
>
  <label
    class="c1 c2"
    for="field"
  >
    label
  </label>
  <div
    class="c3 c4"
  >
    <input
      class="c5 c6"
      disabled=""
      id="field"
      type="text"
      value=""
    />
  </div>
</div>
`
